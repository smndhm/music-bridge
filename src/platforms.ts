interface Platform {
  name: string;
  url: string;
}

interface Platforms {
  [key: string]: Platform;
}

export const platforms: Platforms = {
  amazonMusic: {
    name: 'Amazon Music',
    url: 'https://music.amazon.com',
  },
  appleMusic: {
    name: 'Apple Music',
    url: 'https://music.apple.com',
  },
  deezer: {
    name: 'Deezer',
    url: 'https://www.deezer.com',
  },
  napster: {
    name: 'Napster',
    url: 'https://play.napster.com',
  },
  pandora: {
    name: 'Pandora',
    url: 'https://www.pandora.com',
  },
  soundcloud: {
    name: 'SoundCloud',
    url: 'https://soundcloud.com',
  },
  spotify: {
    name: 'Spotify',
    url: 'https://open.spotify.com',
  },
  tidal: {
    name: 'Tidal',
    url: 'https://listen.tidal.com',
  },
  youtube: {
    name: 'Youtube',
    url: 'https://www.youtube.com',
  },
  youtubeMusic: {
    name: 'Youtube Music',
    url: 'https://music.youtube.com',
  },
};

// https://www.notion.so/Redirect-URLs-9e906d3277d945e29ce422e461fb3c06
// yandex
// spinrilla
// audius
// itunes
// googleStore
// amazonStore

// https://www.notion.so/API-d0ebe08a5e304a55928405eb682f6741
// itunes
// google
// googleStore
// amazonStore
// yandex
// spinrilla
// audius
// audiomack
// anghami
// boomplay
