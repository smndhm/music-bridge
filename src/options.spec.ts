import { describe, it, vi, afterEach, beforeEach, expect } from 'vitest';
import { platforms } from './platforms';
import { readFileSync } from 'fs';
import { JSDOM } from 'jsdom';
import { join } from 'path';

const htmlFilePath = join(__dirname, './options.html');
const html = readFileSync(htmlFilePath, 'utf-8');

// Mock local storage
let platformMock: string;
let openInAppMock: boolean;
let browserStorageLocalSet = vi.fn();

vi.mock('webextension-polyfill', () => ({
  default: {
    storage: {
      local: {
        get: vi.fn(() =>
          Promise.resolve({
            platform: platformMock,
            openInApp: openInAppMock,
          }),
        ),
        set: browserStorageLocalSet,
      },
    },
    i18n: {
      getMessage: vi.fn((key) => key),
    },
  },
}));

/**
 * @vitest-environment jsdom
 */

describe('options', () => {
  beforeEach(() => {
    vi.useFakeTimers();
    vi.resetModules();
    const { window: jsdomWindow } = new JSDOM(html);
    window = jsdomWindow;
    document = window.document;
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it.each(Object.entries(platforms))(
    'Platform (%s)',
    // Run the code
    async (platform) => {
      platformMock = platform;
      await import('./options');
      expect(document.documentElement.outerHTML).toMatchSnapshot();
    },
  );

  it('Should change platform and unselect openDesktopApp option and save', async () => {
    await import('./options');

    const checkbox: HTMLInputElement = document.getElementById(
      'open-app',
    ) as HTMLInputElement;
    checkbox.checked = true;

    const select: HTMLInputElement = document.getElementById(
      'streaming-platforms',
    ) as HTMLInputElement;
    select.value = Object.keys(platforms)[0];
    select.dispatchEvent(new Event('change'));

    expect(checkbox.checked).toEqual(false);

    const form: HTMLFormElement = document.querySelector('form')!;
    const alert: HTMLDivElement = document.querySelector('[role="alert"]')!;

    form.dispatchEvent(new Event('submit'));

    expect(browserStorageLocalSet).toHaveBeenCalledWith({
      platform: select.value,
      openInApp: checkbox.checked,
    });

    await vi.runAllTimersAsync();

    expect(alert.hidden).toEqual(true);
  });
});
