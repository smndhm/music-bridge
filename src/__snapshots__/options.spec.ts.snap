// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`options > Platform (amazonMusic) 1`] = `
"<html lang="@@ui_locale"><head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>optionsTitle</title></head>
  <body>
    <main>
      <form>
        <label for="streaming-platforms">optionsSteamingPlatformLabel
          <select id="streaming-platforms" name="streaming-platforms"><option value="amazonMusic" selected="">Amazon Music</option><option value="appleMusic">Apple Music</option><option value="deezer">Deezer</option><option value="napster">Napster</option><option value="pandora">Pandora</option><option value="soundcloud">SoundCloud</option><option value="spotify">Spotify</option><option value="tidal">Tidal</option><option value="youtube">Youtube</option><option value="youtubeMusic">Youtube Music</option></select>
        </label>
        <label for="open-app" hidden="">
          <input type="checkbox" name="open-app" id="open-app">
        optionsOpenAppLabel</label>
        <div id="save-options">
          <div role="alert" hidden="">optionsSaveConfirmation</div>
          <button>optionsSaveButton</button>
        </div>
      </form>
    </main>
    <footer>optionsCredits</footer>
    <script type="module" src="./options.ts"></script>
  

</body></html>"
`;

exports[`options > Platform (appleMusic) 1`] = `
"<html lang="@@ui_locale"><head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>optionsTitle</title></head>
  <body>
    <main>
      <form>
        <label for="streaming-platforms">optionsSteamingPlatformLabel
          <select id="streaming-platforms" name="streaming-platforms"><option value="amazonMusic">Amazon Music</option><option value="appleMusic" selected="">Apple Music</option><option value="deezer">Deezer</option><option value="napster">Napster</option><option value="pandora">Pandora</option><option value="soundcloud">SoundCloud</option><option value="spotify">Spotify</option><option value="tidal">Tidal</option><option value="youtube">Youtube</option><option value="youtubeMusic">Youtube Music</option></select>
        </label>
        <label for="open-app">
          <input type="checkbox" name="open-app" id="open-app">
        optionsOpenAppLabel</label>
        <div id="save-options">
          <div role="alert" hidden="">optionsSaveConfirmation</div>
          <button>optionsSaveButton</button>
        </div>
      </form>
    </main>
    <footer>optionsCredits</footer>
    <script type="module" src="./options.ts"></script>
  

</body></html>"
`;

exports[`options > Platform (deezer) 1`] = `
"<html lang="@@ui_locale"><head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>optionsTitle</title></head>
  <body>
    <main>
      <form>
        <label for="streaming-platforms">optionsSteamingPlatformLabel
          <select id="streaming-platforms" name="streaming-platforms"><option value="amazonMusic">Amazon Music</option><option value="appleMusic">Apple Music</option><option value="deezer" selected="">Deezer</option><option value="napster">Napster</option><option value="pandora">Pandora</option><option value="soundcloud">SoundCloud</option><option value="spotify">Spotify</option><option value="tidal">Tidal</option><option value="youtube">Youtube</option><option value="youtubeMusic">Youtube Music</option></select>
        </label>
        <label for="open-app">
          <input type="checkbox" name="open-app" id="open-app">
        optionsOpenAppLabel</label>
        <div id="save-options">
          <div role="alert" hidden="">optionsSaveConfirmation</div>
          <button>optionsSaveButton</button>
        </div>
      </form>
    </main>
    <footer>optionsCredits</footer>
    <script type="module" src="./options.ts"></script>
  

</body></html>"
`;

exports[`options > Platform (napster) 1`] = `
"<html lang="@@ui_locale"><head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>optionsTitle</title></head>
  <body>
    <main>
      <form>
        <label for="streaming-platforms">optionsSteamingPlatformLabel
          <select id="streaming-platforms" name="streaming-platforms"><option value="amazonMusic">Amazon Music</option><option value="appleMusic">Apple Music</option><option value="deezer">Deezer</option><option value="napster" selected="">Napster</option><option value="pandora">Pandora</option><option value="soundcloud">SoundCloud</option><option value="spotify">Spotify</option><option value="tidal">Tidal</option><option value="youtube">Youtube</option><option value="youtubeMusic">Youtube Music</option></select>
        </label>
        <label for="open-app" hidden="">
          <input type="checkbox" name="open-app" id="open-app">
        optionsOpenAppLabel</label>
        <div id="save-options">
          <div role="alert" hidden="">optionsSaveConfirmation</div>
          <button>optionsSaveButton</button>
        </div>
      </form>
    </main>
    <footer>optionsCredits</footer>
    <script type="module" src="./options.ts"></script>
  

</body></html>"
`;

exports[`options > Platform (pandora) 1`] = `
"<html lang="@@ui_locale"><head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>optionsTitle</title></head>
  <body>
    <main>
      <form>
        <label for="streaming-platforms">optionsSteamingPlatformLabel
          <select id="streaming-platforms" name="streaming-platforms"><option value="amazonMusic">Amazon Music</option><option value="appleMusic">Apple Music</option><option value="deezer">Deezer</option><option value="napster">Napster</option><option value="pandora" selected="">Pandora</option><option value="soundcloud">SoundCloud</option><option value="spotify">Spotify</option><option value="tidal">Tidal</option><option value="youtube">Youtube</option><option value="youtubeMusic">Youtube Music</option></select>
        </label>
        <label for="open-app" hidden="">
          <input type="checkbox" name="open-app" id="open-app">
        optionsOpenAppLabel</label>
        <div id="save-options">
          <div role="alert" hidden="">optionsSaveConfirmation</div>
          <button>optionsSaveButton</button>
        </div>
      </form>
    </main>
    <footer>optionsCredits</footer>
    <script type="module" src="./options.ts"></script>
  

</body></html>"
`;

exports[`options > Platform (soundcloud) 1`] = `
"<html lang="@@ui_locale"><head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>optionsTitle</title></head>
  <body>
    <main>
      <form>
        <label for="streaming-platforms">optionsSteamingPlatformLabel
          <select id="streaming-platforms" name="streaming-platforms"><option value="amazonMusic">Amazon Music</option><option value="appleMusic">Apple Music</option><option value="deezer">Deezer</option><option value="napster">Napster</option><option value="pandora">Pandora</option><option value="soundcloud" selected="">SoundCloud</option><option value="spotify">Spotify</option><option value="tidal">Tidal</option><option value="youtube">Youtube</option><option value="youtubeMusic">Youtube Music</option></select>
        </label>
        <label for="open-app" hidden="">
          <input type="checkbox" name="open-app" id="open-app">
        optionsOpenAppLabel</label>
        <div id="save-options">
          <div role="alert" hidden="">optionsSaveConfirmation</div>
          <button>optionsSaveButton</button>
        </div>
      </form>
    </main>
    <footer>optionsCredits</footer>
    <script type="module" src="./options.ts"></script>
  

</body></html>"
`;

exports[`options > Platform (spotify) 1`] = `
"<html lang="@@ui_locale"><head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>optionsTitle</title></head>
  <body>
    <main>
      <form>
        <label for="streaming-platforms">optionsSteamingPlatformLabel
          <select id="streaming-platforms" name="streaming-platforms"><option value="amazonMusic">Amazon Music</option><option value="appleMusic">Apple Music</option><option value="deezer">Deezer</option><option value="napster">Napster</option><option value="pandora">Pandora</option><option value="soundcloud">SoundCloud</option><option value="spotify" selected="">Spotify</option><option value="tidal">Tidal</option><option value="youtube">Youtube</option><option value="youtubeMusic">Youtube Music</option></select>
        </label>
        <label for="open-app">
          <input type="checkbox" name="open-app" id="open-app">
        optionsOpenAppLabel</label>
        <div id="save-options">
          <div role="alert" hidden="">optionsSaveConfirmation</div>
          <button>optionsSaveButton</button>
        </div>
      </form>
    </main>
    <footer>optionsCredits</footer>
    <script type="module" src="./options.ts"></script>
  

</body></html>"
`;

exports[`options > Platform (tidal) 1`] = `
"<html lang="@@ui_locale"><head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>optionsTitle</title></head>
  <body>
    <main>
      <form>
        <label for="streaming-platforms">optionsSteamingPlatformLabel
          <select id="streaming-platforms" name="streaming-platforms"><option value="amazonMusic">Amazon Music</option><option value="appleMusic">Apple Music</option><option value="deezer">Deezer</option><option value="napster">Napster</option><option value="pandora">Pandora</option><option value="soundcloud">SoundCloud</option><option value="spotify">Spotify</option><option value="tidal" selected="">Tidal</option><option value="youtube">Youtube</option><option value="youtubeMusic">Youtube Music</option></select>
        </label>
        <label for="open-app" hidden="">
          <input type="checkbox" name="open-app" id="open-app">
        optionsOpenAppLabel</label>
        <div id="save-options">
          <div role="alert" hidden="">optionsSaveConfirmation</div>
          <button>optionsSaveButton</button>
        </div>
      </form>
    </main>
    <footer>optionsCredits</footer>
    <script type="module" src="./options.ts"></script>
  

</body></html>"
`;

exports[`options > Platform (youtube) 1`] = `
"<html lang="@@ui_locale"><head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>optionsTitle</title></head>
  <body>
    <main>
      <form>
        <label for="streaming-platforms">optionsSteamingPlatformLabel
          <select id="streaming-platforms" name="streaming-platforms"><option value="amazonMusic">Amazon Music</option><option value="appleMusic">Apple Music</option><option value="deezer">Deezer</option><option value="napster">Napster</option><option value="pandora">Pandora</option><option value="soundcloud">SoundCloud</option><option value="spotify">Spotify</option><option value="tidal">Tidal</option><option value="youtube" selected="">Youtube</option><option value="youtubeMusic">Youtube Music</option></select>
        </label>
        <label for="open-app" hidden="">
          <input type="checkbox" name="open-app" id="open-app">
        optionsOpenAppLabel</label>
        <div id="save-options">
          <div role="alert" hidden="">optionsSaveConfirmation</div>
          <button>optionsSaveButton</button>
        </div>
      </form>
    </main>
    <footer>optionsCredits</footer>
    <script type="module" src="./options.ts"></script>
  

</body></html>"
`;

exports[`options > Platform (youtubeMusic) 1`] = `
"<html lang="@@ui_locale"><head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>optionsTitle</title></head>
  <body>
    <main>
      <form>
        <label for="streaming-platforms">optionsSteamingPlatformLabel
          <select id="streaming-platforms" name="streaming-platforms"><option value="amazonMusic">Amazon Music</option><option value="appleMusic">Apple Music</option><option value="deezer">Deezer</option><option value="napster">Napster</option><option value="pandora">Pandora</option><option value="soundcloud">SoundCloud</option><option value="spotify">Spotify</option><option value="tidal">Tidal</option><option value="youtube">Youtube</option><option value="youtubeMusic" selected="">Youtube Music</option></select>
        </label>
        <label for="open-app" hidden="">
          <input type="checkbox" name="open-app" id="open-app">
        optionsOpenAppLabel</label>
        <div id="save-options">
          <div role="alert" hidden="">optionsSaveConfirmation</div>
          <button>optionsSaveButton</button>
        </div>
      </form>
    </main>
    <footer>optionsCredits</footer>
    <script type="module" src="./options.ts"></script>
  

</body></html>"
`;
